grouped_points[] grouped_points_array
